modules/core/04-channel/keeper/upgrade.go

```diff
@@ -22,115 +14,64 @@ func (k Keeper) ChanUpgradeInit(
+		return proposedUpgrade, errorsmod.Wrapf(types.ErrInvalidChannelState, "expected %s, got %s", types.OPEN, channel.State)
 	}
 
-	// set the restore channel to the current channel and reassign channel state to INITUPGRADE,
-	// if the channel == proposedUpgradeChannel then fail fast as no upgradable fields have been modified.
-	restoreChannel := channel
-	channel.State = types.INITUPGRADE
-	if reflect.DeepEqual(channel, proposedUpgradeChannel) {
-		return 0, "", errorsmod.Wrap(types.ErrChannelExists, "existing channel end is identical to proposed upgrade channel end")
+	if err := k.ValidateUpgradeFields(ctx, upgradeFields, channel); err != nil {
+		return proposedUpgrade, err
 	}
 
-	connectionEnd, err := k.GetConnection(ctx, proposedUpgradeChannel.ConnectionHops[0])
+	proposedUpgrade, err = k.constructProposedUpgrade(ctx, portID, channelID, upgradeFields, upgradeTimeout)
 	if err != nil {
-		return 0, "", err
-	}
-
-	if connectionEnd.GetState() != int32(connectiontypes.OPEN) {
-		return 0, "", errorsmod.Wrapf(
-			connectiontypes.ErrInvalidConnectionState,
-			"connection state is not OPEN (got %s)", connectiontypes.State(connectionEnd.GetState()).String(),
-		)
-	}
-
-	if proposedUpgradeChannel.Counterparty.PortId != channel.Counterparty.PortId ||
-		proposedUpgradeChannel.Counterparty.ChannelId != channel.Counterparty.ChannelId {
-		return 0, "", errorsmod.Wrap(types.ErrInvalidCounterparty, "counterparty port ID and channel ID cannot be upgraded")
+		return proposedUpgrade, errorsmod.Wrap(err, "failed to construct proposed upgrade")
 	}
 
-	if !proposedUpgradeChannel.Ordering.SubsetOf(channel.Ordering) {
-		return 0, "", errorsmod.Wrap(types.ErrInvalidChannelOrdering, "channel ordering must be a subset of the new ordering")
-	}
-
-	upgradeSequence = uint64(1)
-	if seq, found := k.GetUpgradeSequence(ctx, portID, channelID); found {
-		upgradeSequence = seq + 1
-	}
-
-	upgradeTimeout := types.UpgradeTimeout{
-		TimeoutHeight:    counterpartyTimeoutHeight,
-		TimeoutTimestamp: counterpartyTimeoutTimestamp,
-	}
-
-	k.SetUpgradeRestoreChannel(ctx, portID, channelID, restoreChannel)
-	k.SetUpgradeSequence(ctx, portID, channelID, upgradeSequence)
-	k.SetUpgradeTimeout(ctx, portID, channelID, upgradeTimeout)
+	channel.UpgradeSequence++
+	k.SetChannel(ctx, portID, channelID, channel)
 
-	return upgradeSequence, channel.Version, nil
+	return proposedUpgrade, nil
 }
 
 // WriteUpgradeInitChannel writes a channel which has successfully passed the UpgradeInit handshake step.
 // An event is emitted for the handshake step.
-func (k Keeper) WriteUpgradeInitChannel(
-	ctx sdk.Context,
-	portID,
-	channelID string,
-	upgradeSequence uint64,
-	channelUpgrade types.Channel,
-) {
+func (k Keeper) WriteUpgradeInitChannel(ctx sdk.Context, portID, channelID string, currentChannel types.Channel, upgrade types.Upgrade) {
```

Comment:
The only thing we have to be careful with about passing in the channel is to ensure it is obtained and not modified in state before calling this function

Reply:
we could do an additional read of the channel directly in the function, I'm not sure it's necessary though, WDYT?

Reply:
It is probably safer to do an additional read but I'm also happy to leave as is for now. @colin-axner and I discussed yesterday briefly about refactoring grpc handlers into their associated submodules rather than having the entry point at the core layer. This would mean we could reduce the exported APIs and prevent them from being misused. If this were a private method then it would be much safer passing `currentChannel`.